File: services/confluence-importer/src/embeddings/openai.ts
import { withRetry } from "../utils/retry.js";
import type { EmbeddingClient, EmbeddingConfig } from "./types.js";

export class OpenAIEmbedding implements EmbeddingClient {
  private readonly apiKey: string;
  readonly model: string;
  readonly dimensions: number;
  private readonly baseUrl = "https://api.openai.com/v1";

  constructor(config: EmbeddingConfig) {
    this.apiKey = config.apiKey;
    this.model = config.model;
    this.dimensions = config.dimensions ?? 1536;
  }

  private static readonly MAX_BATCH = 2048;

  async embed(texts: string[]): Promise<number[][]> {
    // Sub-batch to stay within the 2048 input limit
    if (texts.length > OpenAIEmbedding.MAX_BATCH) {
      const results: number[][] = [];
      for (let i = 0; i < texts.length; i += OpenAIEmbedding.MAX_BATCH) {
        const batch = texts.slice(i, i + OpenAIEmbedding.MAX_BATCH);
        const batchResults = await this.embed(batch);
        results.push(...batchResults);
      }
      return results;
    }

    return withRetry(async () => {
      const response = await fetch(`${this.baseUrl}/embeddings`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${this.apiKey}`,
        },
        body: JSON.stringify({
          model: this.model,
          input: texts,
          dimensions: this.dimensions,
        }),
      });

      if (!response.ok) {
        const error = await response.text();
        throw new Error(`OpenAI embedding failed (${response.status}): ${error}`);
      }

      const data = (await response.json()) as {
        data: Array<{ index: number; embedding: number[] }>;
      };
      const sorted = [...data.data].sort((a, b) => a.index - b.index);
      return sorted.map((item) => item.embedding);
    });
  }

  async embedSingle(text: string): Promise<number[]> {
    const [embedding] = await this.embed([text]);
    return embedding;
  }
}

File: services/confluence-importer/src/confluence_importer/embeddings/openai.py
"""OpenAI embeddings via the REST API (no SDK dependency)."""

import httpx

from confluence_importer.embeddings.base import EmbeddingApiError, batched, validate_dimensions
from confluence_importer.retry import with_retry

_BASE_URL = "https://api.openai.com/v1"
_MAX_BATCH = 100


class OpenAIEmbedding:
    def __init__(
        self,
        *,
        api_key: str,
        model: str,
        dimensions: int = 1536,
        http_client: httpx.Client | None = None,
    ) -> None:
        self.model = model
        self.dimensions = dimensions
        self._http = http_client or httpx.Client(
            headers={"Authorization": f"Bearer {api_key}"},
            timeout=60.0,
        )

    def embed_texts(self, texts: list[str]) -> list[list[float]]:
        results: list[list[float]] = []
        for batch in batched(texts, _MAX_BATCH):
            results.extend(self._embed_batch(batch))
        return results

    def embed_query(self, text: str) -> list[float]:
        return self.embed_texts([text])[0]

    def _embed_batch(self, texts: list[str]) -> list[list[float]]:
        def do_request() -> list[list[float]]:
            response = self._http.post(
                f"{_BASE_URL}/embeddings",
                json={"model": self.model, "input": texts, "dimensions": self.dimensions},
            )
            if response.status_code >= 400:
                raise EmbeddingApiError("OpenAI", response.status_code, response.text)
            data = sorted(response.json()["data"], key=lambda item: item["index"])
            embeddings = [item["embedding"] for item in data]
            return validate_dimensions("OpenAI", self.model, self.dimensions, embeddings)

        return with_retry(do_request)
